File: src/algorithm/Sweeping.java
package algorithm;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class Sweeping {
    
    public static class Segment {
        int start;
        int end;

        Segment(int start, int end) {
            this.start = start;
            this.end = end;
        }
    }

    public static class Event {
        int point;
        String eventType;

        Event(int point, String eventType) {
            this.point = point;
            this.eventType = eventType;
        }
    }

    public static void main(String[] args) {
        List<Segment> segments = new ArrayList<>();
        segments.add(new Segment(1, 4));
        segments.add(new Segment(2, 5));
        segments.add(new Segment(3, 6));
        segments.add(new Segment(4, 7));
        segments.add(new Segment(5, 8));

        int result = countIntersectingSegments(segments);
        System.out.println("Number of intersecting segments: " + result);
    }

    public static int countIntersectingSegments(List<Segment> segments) {
        List<Event> events = new ArrayList<>();
        for (Segment segment : segments) {
            events.add(new Event(segment.start, "start"));
            events.add(new Event(segment.end, "end"));
        }

        Collections.sort(events, Comparator.comparingInt(e -> e.point)); // 오름차순 정렬

        int count = 0;
        int intersecting = 0;
        for (Event event : events) {
            if (event.eventType.equals("start")) {
                intersecting++;
            } else if (event.eventType.equals("end")) {
                intersecting--;
            }
            count = Math.max(count, intersecting);
        }

        return count;
    }


}

File: bin/algorithm/Sweeping.py
# 스위프(sweep)라는 가상의 수직선 또는 수평선을 사용하여 공간을 효과적으로 탐색하는 방식
# 겹치는 선분의 개수를 세는 스위핑 알고리즘

def count_intersecting_segments(segments):
    events = []
    for start, end in segments:
        events.append((start, "start"))
        events.append((end, "end"))

    events.sort()  # 이벤트 리스트를 정렬

    count = 0
    intersecting = 0
    for point, event_type in events:
        if event_type == "start":
            intersecting += 1
        elif event_type == "end":
            intersecting -= 1
        count = max(count, intersecting)

    return count


segments = [(1, 4), (2, 5), (3, 6), (4, 7), (5, 8)]
result = count_intersecting_segments(segments)
print("Number of intersecting segments:", result)